Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,905 @@
+    newK a τ x = (g ^ m) (a ((τ  : Perm g.cycleFactorsFinset) c)) := by
+  have hx' : c = g.cycleOf x := cycle_is_cycleOf hx (Subtype.prop c)
+  have hx'' : g.cycleOf x ∈ g.cycleFactorsFinset := hx' ▸ c.prop
+  let h := mem_support_cycleOf_iff.mp (a.mem_support_self ⟨g.cycleOf x, hx''⟩)
+  set n := h.1.symm.choose
+  have hn : (g ^ n) (a c) = x := by
+    rw [← h.1.symm.choose_spec]
+    congr
+    rw [← Subtype.coe_inj, hx']
+  suffices newK a τ x = (g ^ n) (a ((τ : Perm g.cycleFactorsFinset) c)) by
+    rw [this, IsCycleOn.zpow_apply_eq_zpow_apply
+      (isCycleOn_support_of_mem_cycleFactorsFinset ((τ : Perm g.cycleFactorsFinset) c).prop)
+      (mem_support_self a ((τ : Perm g.cycleFactorsFinset) c))]
+    simp only [τ.prop c]
+    rw [← IsCycleOn.zpow_apply_eq_zpow_apply
+      (isCycleOn_support_of_mem_cycleFactorsFinset c.prop) (mem_support_self a c)]
+    rw [hn, hm]
+  simp only [newK, dif_pos hx'']
+  congr
+  exact hx'.symm
+
+theorem newK_apply_of_mem_fixedPoints {x : α} (hx : x ∈ Function.fixedPoints g) :
+    newK a τ x = x := by
+  rw [newK, dif_neg]
+  rw [cycleOf_mem_cycleFactorsFinset_iff, not_mem_support]
+  exact hx
+
+theorem zpow_apply_mem_support_of_mem_cycleFactorsFinset_iff
+    {x :α} {m : ℤ} {c : g.cycleFactorsFinset} :
+    (g ^ m) x ∈ (c : Perm α).support ↔ x ∈ (c : Perm α).support := by
+  rw [← g.eq_cycleOf_of_mem_cycleFactorsFinset_iff _ c.prop]
+  rw [cycleOf_self_apply_zpow g m x]
+  rw [eq_cycleOf_of_mem_cycleFactorsFinset_iff _ _ c.prop]
+
+theorem newK_apply_mem_support_cycle_iff {x : α} {c : g.cycleFactorsFinset} :
+    newK a τ x ∈ ((τ : Perm g.cycleFactorsFinset) c : Perm α).support ↔
+      x ∈ (c : Perm α).support := by
+  rcases mem_fixedPoints_or_exists_zpow_eq a x with (hx | ⟨d, hd, m, hm⟩)
+  · simp only [newK_apply_of_mem_fixedPoints a τ hx]
+    suffices ∀ (d : g.cycleFactorsFinset), x ∉ (d : Perm α).support by
+      simp only [this]
+    intro d hx'
+    rw [Function.mem_fixedPoints_iff, ← not_mem_support] at hx
+    apply hx
+    exact mem_cycleFactorsFinset_support_le d.prop hx'
+  · rw [newK_apply_of_cycleOf_mem a τ hd hm] --
+    rw [zpow_apply_mem_support_of_mem_cycleFactorsFinset_iff]
+    by_cases h : c = d
+    · simp only [h, hd, iff_true, mem_support_self]
+    · have H : Disjoint (c : Perm α) (d : Perm α) :=
+        cycleFactorsFinset_pairwise_disjoint g c.prop d.prop (Subtype.coe_ne_coe.mpr h)
+      have H' : Disjoint ((τ : Perm g.cycleFactorsFinset) c : Perm α)
+        ((τ : Perm g.cycleFactorsFinset) d : Perm α) :=
+        cycleFactorsFinset_pairwise_disjoint g ((τ : Perm g.cycleFactorsFinset) c).prop
+          ((τ : Perm g.cycleFactorsFinset) d).prop (by
+          intro h'; apply h
+          simpa only [Subtype.coe_inj, EmbeddingLike.apply_eq_iff_eq] using h')
+      rw [disjoint_iff_disjoint_support, Finset.disjoint_right] at H H'
+      simp only [H hd, H' (mem_support_self a _)]
+
+theorem newK_apply_mem_fixedPoints_iff {x : α} :
```

Comment:
I don't think this is used?

Reply:
Yes it is.

Reply:
But it's the iff version of something that is used and it's better to have it, I'd say.

Reply:
anyway, now it's deleted.